crates/teloxide/src/dispatching/dialogue/get_chat_id.rs

```diff
@@ -23,3 +26,89 @@ impl GetChatId for Update {
         self.chat().map(|chat| chat.id)
     }
 }
+
+impl GetChatId for Recipient {
+    fn chat_id(&self) -> Option<ChatId> {
+        match self {
+            Recipient::Id(chat_id) => Some(*chat_id),
+            Recipient::ChannelUsername(_) => None,
+        }
+    }
+}
+
+impl GetChatId for BotCommandScope {
+    fn chat_id(&self) -> Option<ChatId> {
+        match self {
+            BotCommandScope::Default
+            | BotCommandScope::AllPrivateChats
+            | BotCommandScope::AllGroupChats
+            | BotCommandScope::AllChatAdministrators => None,
+            BotCommandScope::Chat { chat_id: recipient }
+            | BotCommandScope::ChatAdministrators { chat_id: recipient }
+            | BotCommandScope::ChatMember { chat_id: recipient, .. } => recipient.chat_id(),
+        }
+    }
+}
+
+impl GetChatId for Chat {
+    fn chat_id(&self) -> Option<ChatId> {
+        Some(self.id)
+    }
+}
+
+impl GetChatId for UpdateKind {
+    fn chat_id(&self) -> Option<ChatId> {
+        match self {
+            UpdateKind::Message(message)
+            | UpdateKind::EditedMessage(message)
+            | UpdateKind::ChannelPost(message)
+            | UpdateKind::EditedChannelPost(message) => GetChatId::chat_id(message),
+            UpdateKind::CallbackQuery(callback_query) => callback_query.chat_id(),
+            UpdateKind::MyChatMember(chat_member_updated) => chat_member_updated.chat_id(),
+            UpdateKind::ChatMember(chat_member_updated) => chat_member_updated.chat_id(),
+            UpdateKind::ChatJoinRequest(chat_join_request) => chat_join_request.chat_id(),
+            UpdateKind::InlineQuery(_)
+            | UpdateKind::ChosenInlineResult(_)
+            | UpdateKind::ShippingQuery(_)
+            | UpdateKind::PreCheckoutQuery(_)
+            | UpdateKind::Poll(_)
+            | UpdateKind::PollAnswer(_)
+            | UpdateKind::Error(_) => None,
+        }
+    }
+}
+
+impl GetChatId for ResponseParameters {
+    fn chat_id(&self) -> Option<ChatId> {
+        match self {
+            ResponseParameters::MigrateToChatId(chat_id) => Some(*chat_id),
+            ResponseParameters::RetryAfter(_) => None,
+        }
+    }
+}
+
+impl GetChatId for TargetMessage {
+    fn chat_id(&self) -> Option<ChatId> {
+        match self {
+            TargetMessage::Common { chat_id: recipient, .. } => recipient.chat_id(),
+            TargetMessage::Inline { .. } => None,
+        }
+    }
+}
+
+/// Implements [`GetChatId`] for all types passed in, as long as they have a
+/// `chat` field with a `Chat` type
+macro_rules! impl_GetChatId_for_chat_field {
```

Comment:
Maybe it would be better to `#[derive(GetChatId)]?

I think, it's possible with macros to check, whether the struct has `chat_id` field, so, we can create "one macro to rule them all"

[Field](https://docs.rs/syn/latest/syn/struct.Field.html)

Reply:
Derive macros require quite a bit of code to implement, I don't think it's useful here. Tbh I don't think this macro is useful either, there are not enough impls to justify it, it would be simpler to just write them out 😅

Reply:
That is true. If we are going forward with this derive macro idea, it would be more beneficial to have it as an attribute macro.